### ISSUE TITLE
The installation takes a very long time and requires a large amount of disk space.

### ISSUE THREAD

**Issue by niklwors** [OpenModelica/OpenModelica#7776]
### Description

The installation of the OpenModelica with the Windwos  setup  file takes a very long time and requires a large amount of disk space. please see attached pictures.

### Screenshots
![discspace](https://user-images.githubusercontent.com/12269776/129692428-1ac14d81-7635-4597-aaad-dd0b21d94cc1.png)
![discspace2](https://user-images.githubusercontent.com/12269776/129692444-55425048-92f0-4d4a-9948-a0f97de6143f.png)
![dsicspace3](https://user-images.githubusercontent.com/12269776/129692476-813036e4-7ab6-4d14-92b3-994049b11f37.png)
![discspace4](https://user-images.githubusercontent.com/12269776/129692504-263676a3-3943-4953-92de-eaa6353a2b12.png)

### Version and OS

 - OpenModelica Version: v19
 - OS: Windows 10, 64 bit
 - Versions of used Modelica libraries if applicable

**Comment by casella** [comment]
@niklwors did you notice a significant worsening with 1.19.0 compared to previous versions, or is this a general comment?

I've been installing OMC once or twice a week; it's a bit cumbersome, but still acceptable on a modern PC with 8 GB or RAM, a fast enough internet connection, and an SSD, which is pretty much standard nowadays.

**Comment by niklwors** [comment]
@casella  My colleagues noticed it, so I created a ticket for it. A lot of what is installed, I think is no longer used.

**Comment by casella** [comment]
> @casella My colleagues noticed it, so I created a ticket for it. 

I'm afraid there's not much we can do about this, the problem is that we have a lot of dependencies on OS libraries (including the whole MinGW thing) and we need to include them in the installer. Linux with APT is a lot smarter

> A lot of what is installed, I think is no longer used.

I'm not aware of anything unused in the installer. Can you be more specific?

**Comment by niklwors** [comment]
> 
> 
> > @casella My colleagues noticed it, so I created a ticket for it.
> 
> I'm afraid there's not much we can do about this, the problem is that we have a lot of dependencies on OS libraries (including the whole MinGW thing) and we need to include them in the installer. Linux with APT is a lot smarter
> 
Yes this is correct the whole MinGW is  important and very good that it is installed.
> > A lot of what is installed, I think is no longer used.
> 
> I'm not aware of anything unused in the installer. Can you be more specific?

For example the header files and libraries for the different runtimes C,C++,omsic, omsicpp. Maybe by default only C and C++ is installed. Or the user can decide which runtime he want to install. Also the libraries for the Visual Studio compiler can be disabled by default. I think there some other libraries which are not used by default users like the optimization or python libraries.

**Comment by casella** [comment]
@niklwors, @adrpo could have a further look at this. Back then, it took him a really long time to finally manage to squeeze everything that was needed for the 64 bit Windows version in the installer, I'm not sure it looking into it again is worth the effort.

Maybe the best solution to solve this issue is to use Windows Subsystem for Linux, instead of MinGW & Co. This would make the installation process via APT much more lightweight and straigtforward

**Comment by casella** [comment]
Another option is to supply a virtual machine with preinstalled OMC. This is much faster to install, though it will probably take even more hard disk space.

**Comment by hkiel** [comment]
Here is an installation time report with the fix in #9068 (39000 files, 347 MB less)
| version | install | uninstall |
| --: | --: | --: |
| 1.19.0 | 500s |  240s |
| 1.20.0-dev-140 | 380s |  160s |

Removing the MSVC boost headers (10800 files, 99 MB) could reduce installation time further.
@adrpo Are they really needed?

Another option would be to remove contents of `tools/msys/mingw64/share/doc` (11800 files, 121 MB)

**Comment by casella** [comment]
@hkiel thanks for #9068, this is really appreciated, as I uninstall and reinstall nightlies on a weekly basis (or more). 

@adrpo can comment on MSVC boost headers, maybe also @mahge.

I'm also in favour of eliminating the doc files, after all the installation is not meant for OMC development (you get stuff from GitHub in that case) and OpenModelica users have no use for those documents. @hkiel please go ahead with that.

**Comment by hkiel** [comment]
@casella I cannot find the place in the Makefiles where OMDev is copied, also the OMDev repository itself (`https://openmodelica.org/git/OMDev.git`) seems to be broken.
```
$ git clone https://openmodelica.org/git/OMDev.git
Cloning into 'OMDev'...
fatal: unable to access 'https://openmodelica.org/git/OMDev.git/': SSL certificate problem: certificate has expired
```

**Comment by adrpo** [comment]
We moved openmodelica.org to another server, I guess we need to configure some thing to get OMDev git working again. I will look into it.

> @casella I cannot find the place in the Makefiles where OMDev is copied, also the OMDev repository itself (`https://openmodelica.org/git/OMDev.git`) seems to be broken.
Because is not in the Makefiles. Is in our NSIS installer:
https://github.com/OpenModelica/OpenModelicaSetup/
At this place:
https://github.com/OpenModelica/OpenModelicaSetup/blob/master/OpenModelicaSetup.nsi#L169

Note that the boost MSVC headers are be different than the MINGW based ones. As far as I know they are needed to build for the msvc target (which I don't know if it still works) for C and C++ runtimes.

**Comment by hkiel** [comment]
> Because is not in the Makefiles. Is in our NSIS installer: https://github.com/OpenModelica/OpenModelicaSetup/ At this place: https://github.com/OpenModelica/OpenModelicaSetup/blob/master/OpenModelicaSetup.nsi#L169

@adrpo Can you add `doc` to the excluded dirs in the installer file?

**Comment by casella** [comment]
I agree with @mahge, there is no point shipping stuff that we are not testing and may be broken.

**Comment by casella** [comment]
@mahge just one thing I'm a bit confused about. It's one thing to build OMC with MSVC, and another thing to use MSVC to compile the simulation executables when you compile a Modelica model with OMC. I understood we were talking about the second option, but maybe you were actually referring to the first.

In that case, I understand those libraries are runtime DLLs needed to run omc.exe or OMEdit.exe if it is built with MSVC. Which is not the case of the shipped OMEdit.exe and omc.exe, so there's no point having them there.

Do I understand correctly?

**Comment by mahge** [comment]
>  I understood we were talking about the second option, but maybe you were actually referring to the first.

You are right (?) the libraries built with MSVC now are not needed by `omc` itslef. They are only for simulation time usage when one selects MSVC as the target compiler. However that is probably broken because it has not been used for a long time as far as I know. If anyone is using it actively then we can keep it.

There is, however, a new CMake build that builds everything (omc itself and all the simulation runtime libraries) with MSVC. If people are actually using MSVC then we can use that build to create the needed libraries and ship them. Either way the current (non-CMake) MSVC built libraries are probably not useful since simulation code compilation with MSVC is probably broken anyway.

**Comment by AnHeuermann** [comment]
OpenModelica (v1.27.0-dev) still is huge and needs around 7.65 GB on my Windows machine. Around 6.47 GB of that is the MSYS environment we ship with OpenModelica. If switching to CMake + MSVC is feasible we can probably use Windows native C/C++ compilers ([Buildtools for Visual Studio 2026](https://visualstudio.microsoft.com/de/downloads/#build-tools-for-visual-studio-2026)) instead and get rid of OMDev. We might need to remove/update some scripting API functions that shell out to bash.

**Comment by hkiel** [comment]
Installation time of v1.20-dev was down to 6min 20s. I now installed v1.26.3 on the same machine and that took 18min 30s!

The current installation is 7.8GB and 126.103 files in 8.380 folders.
I just looked up the installer file sizes and that goes up during the last releases. Notably from 1.22.3 to 1.23.0.

| version | size
------|-------
v1.22.3 | 1.4GB
v1.23.0 | 1.7GB
v1.24.0 | 1.8GB
v1.24.5 | 1.9GB
v1.25.0 | 2.0GB

in 1.26.x the compression algorithm was adjusted to go to 1.6GB again.

I doubt that all includes and libs in the msys directory are needed to compile Modelica models with omc/OMEdit.
E.g. there is `aws` and `vtk`. Also `python3.11` and `perl5` bring a huge amount of files.

**Comment by cbu-hw** [comment]
> If switching to CMake + MSVC is feasible we can probably use Windows native C/C++ compilers ([Buildtools for Visual Studio 2026](https://visualstudio.microsoft.com/de/downloads/#build-tools-for-visual-studio-2026)) instead and get rid of OMDev. We might need to remove/update some scripting API functions that shell out to bash.

Is that feasible/desirable, to move away from an open toolchain? Last time I checked, the MSVC build tools needed an active Visual Studio license in many (e.g. corporate) contexts.

**Comment by hkiel** [comment]
> > If switching to CMake + MSVC is feasible we can probably use Windows native C/C++ compilers ([Buildtools for Visual Studio 2026](https://visualstudio.microsoft.com/de/downloads/#build-tools-for-visual-studio-2026)) instead and get rid of OMDev. We might need to remove/update some scripting API functions that shell out to bash.
> 
> Is that feasible/desirable, to move away from an open toolchain? Last time I checked, the MSVC build tools needed an active Visual Studio license in many (e.g. corporate) contexts.

@casella also stated some time ago that with every new release of MSVC you get new incompatibilities and problems.

**Comment by adeas31** [comment]
We should have different installer for OMDev. We hardly change anything in OMDev. So we should install OpenModelica and OMDev separately.

I had that in mind for long but didn't get enough time to fix it. Maybe its time to prioritize this after 1.27 release.

**Comment by AnHeuermann** [comment]
> Is that feasible/desirable, to move away from an open toolchain? Last time I checked, the MSVC build tools needed an active Visual Studio license in many (e.g. corporate) contexts.

Yeah, probably correct. In theory any OpenModelica user could use the MSVC tools to compile arbitrary C/C++ code. That's for sure not possible without proper licensing.

**Comment by AnHeuermann** [comment]
According to @adrpo the next version of OMDev will be even larger. Fine for developing but for the installer we need to slim it down to the bare minimum and re-use existing MSYS whenever possible like @adeas31 suggested.

**Comment by hkiel** [comment]
> According to [@adrpo](https://github.com/adrpo) the next version of OMDev will be even larger. Fine for developing but for the installer we need to slim it down to the bare minimum and re-use existing MSYS whenever possible like [@adeas31](https://github.com/adeas31) suggested.

As far as I remember there were two versions of OMDev: one for the end-user, which includes everything to **run** OM, and the other for OM developers.

**Comment by adrpo** [comment]
The one in the installer is heavily filtered from the developer one but I think it can be even further shrinked. I think python is needed for gdb that we run when OMEdit crashes. I am not sure about pearl, I will check.

**Comment by adrpo** [comment]
What takes a long time is keeping track of all the files. The installer since 1.26 will build a list of all the files it installs so that it will only delete that on uninstall. We had issues where people installed OM in C:\ and the uninstall wiped their files.

**Comment by hkiel** [comment]
> What takes a long time is keeping track of all the files. The installer since 1.26 will build a list of all the files it installs so that it will only delete that on uninstall. We had issues where people installed OM in C:\ and the uninstall wiped their files.

That's only half of the truth. The other big problem is that windows does not work well with directories containing more than ~500 files (which holds true for `include/` and other directories in MSYS). Any file operation will slow when there are "too many" files in a directory. That seems to be a limitation of NTFS/FAT.

**Comment by hkiel** [comment]
> What takes a long time is keeping track of all the files. The installer since 1.26 will build a list of all the files it installs so that it will only delete that on uninstall. We had issues where people installed OM in C:\ and the uninstall wiped their files.

@adrpo Could a compromise be that only top level files and directories are stored in that file and the dirs be deleted recursively?

**Comment by perost** [comment]
> > What takes a long time is keeping track of all the files. The installer since 1.26 will build a list of all the files it installs so that it will only delete that on uninstall. We had issues where people installed OM in C:\ and the uninstall wiped their files.
> 
> [@adrpo](https://github.com/adrpo) Could a compromise be that only top level files and directories are stored in that file and the dirs be deleted recursively?

The issue with that is that some users might decide to store for example their models in the OM directory, I have a vague memory of getting some reports about that in the past. So ideally we shouldn't touch any files we haven't installed.

Of course, if it takes more time to make a list of the installed files than it takes to actually write them to disk, then something is very wrong somewhere.

**Comment by hkiel** [comment]
> The issue with that is that some users might decide to store for example their models in the OM directory, I have a vague memory of getting some reports about that in the past. So ideally we shouldn't touch any files we haven't installed.

If default paths for the OM tools are outside MSYS directory we could at least remove the whole MSYS directory, which is the biggest part of the (windows) installation. I guess nobody explicitly choses that dir for models.

As far as I remember `cd()` in OMNotebook input-cell initially points to `$tmp` directory, which is also bad as some OSes regularly clean `$tmp` (e.g. at reboot).
